bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/discovery/HomeAssistantDiscovery.java

```diff
@@ -242,24 +263,25 @@ private DiscoveryResult buildResult(String thingID, ThingUID thingUID, String th
                 .withBridge(bridgeUID).withLabel(thingName).build();
     }
 
+    /**
+     * Stores the newly built DiscoveryResult and marks it dirty.
+     */
     private void applyResult(String thingID, HaID haID, DiscoveryResult result) {
-        // Mutations only — must be under discoveryStateLock
-        Set<HaID> componentsUnordered = componentsPerThingID.computeIfAbsent(thingID, key -> new HashSet<>());
-        componentsUnordered.add(haID);
-
-        results.put(result.getThingUID().toString(), result);
         allResults.put(result.getThingUID().toString(), result);
+        dirtyResults.add(result.getThingUID());
     }
 
     protected void publishResults() {
-        Collection<DiscoveryResult> localResults;
-
+        Set<ThingUID> toPublish = new HashSet<>();
```

Comment:
wouldn't it even less allocation to do this:

```java
Set<ThingUID> toPublish;
 synchronized (discoveryStateLock) {
  toPublish = dirtyResults;
  dirtyResults = new HashSet<>();
}
```

that way you're allocating a new set (which we're doing anyway), but _not_ having to add each item to that set.

Reply:
There are some pro's and cons. The performance impact depends on the size of the set. 

both allocate a new HashSet each call, in both cases, a set instance becomes garbage after use.

The difference is that my implementation clears the set and keeps its allocation. Yours does a pointer swap. 
In the end it comes down to the size of the set to determine best performance. mine = for small sets, yours is for large sets.

Atleast that is my analysis, i might be wrong somewhere. Let me know i fyou want me to adapt it.

Reply:
What about a hybrid solution? Fastest, avoids both copying and repeated allocations.

private Set<ThingUID> spareResults = new HashSet<>();
    synchronized (discoveryStateLock) {
        // Swap dirty and spare
        toPublish = dirtyResults;
        dirtyResults = spareResults;
        spareResults = toPublish;

        // Make sure the new dirty set is empty
        dirtyResults.clear();
    }